### PR TITLE
[HIP] Adds support for creating devices

### PR DESCRIPTION
**Pull request by nithinsubbiah**
Progress towards #15790

### CODE REVIEW

File: experimental/hip/hip_device.c [hunk 1]
```diff
@@ -0,0 +1,513 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "experimental/hip/hip_device.h"
+
+#include <iree/hal/device.h>
```
Comment (antiagainst) [inline experimental/hip/hip_device.c]: This include is not needed.

File: experimental/hip/hip_device.c [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+    iree_hal_driver_t* driver, iree_string_view_t identifier,
+    const iree_hal_hip_device_params_t* params, hipDevice_t hip_device,
+    hipStream_t stream, hipCtx_t context,
+    const iree_hal_hip_dynamic_symbols_t* symbols,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  iree_hal_hip_device_t* device = NULL;
+  iree_host_size_t total_size = iree_sizeof_struct(*device) + identifier.size;
+  IREE_RETURN_IF_ERROR(
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device));
+  memset(device, 0, total_size);
```
Comment (antiagainst) [inline experimental/hip/hip_device.c]: This is not needed actually--if you look at `iree_allocator_malloc`, it calls into the `CALLOC` variant, which will zero the allocation along the way. Feel free to fix the cuda2 side too.

File: experimental/hip/hip_device.c [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+
+  // Get the main context for the device.
+  hipCtx_t context = NULL;
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols, hipDevicePrimaryCtxRetain(&context, device));
+  }
+  if (iree_status_is_ok(status)) {
+    status = IREE_HIP_RESULT_TO_STATUS(
+        symbols, hipSetDevice(IREE_HIPDEVICE_TO_DEVICE_ID(device)));
```
Comment (antiagainst) [inline experimental/hip/hip_device.c]: From the [API](https://docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___device.html#ga43c1e7f15925eeb762195ccb5e063eae), `hipSetDevice` expects an `int` device index right? `hipDevice_t` is queried from `hipGetDevice` with a device index. Doc doesn't say `hipDevice_t` is just device index under the hood; so we cannot assume so. (Better to ask internally regarding this.) Therefore we may additionally need to keep track of the given device's index in its struct. Also, `IREE_HIPDEVICE_TO_DEVICE_ID` is not necessary here given that converts the `device` to an ID used in IREE.

File: experimental/hip/hip_device.c [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_hip_device_create_internal(
+        driver, identifier, params, device, stream, context, symbols,
+        host_allocator, out_device);
+  }
+  if (!iree_status_is_ok(status)) {
+    if (stream) symbols->hipStreamDestroy(stream);
+    // WARNING: This function return hipSuccess though doesn't release the
```
Comment (antiagainst) [inline experimental/hip/hip_device.c]: Change WARNING to Note would be fine.

File: experimental/hip/hip_device.c [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+  iree_hal_allocator_release(device->device_allocator);
+
+  // Destroy memory pools that hold on to reserved memory.
+  iree_hal_hip_memory_pools_deinitialize(&device->memory_pools);
+
+  // TODO: support multiple streams.
+  IREE_HIP_IGNORE_ERROR(device->hip_symbols,
+                        hipStreamDestroy(device->hip_stream));
+
+  // WARNING: This function return hipSuccess though doesn't release the
```
Comment (antiagainst) [inline experimental/hip/hip_device.c]: Similarly here.

**Review by antiagainst — CHANGES_REQUESTED** [review]

**Comment by antiagainst** [comment]
The lastest API doc marks the majority of the context management APIs as [deprecated](https://rocm.docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___context_d.html#ga8aa32cf64272da929f23ecbafefefcee), with some exceptions like `hipDevicePrimaryCtxRetain`. Not sure that's intentional or just a bug in the doc. Could you ask internally to clarify regarding this aspect? And also regarding primary context management (is it really needed with `hipDevicePrimaryCtxRelease` doing nothing?) And what is the expected way to manage context in general, with a bunch of APIs deprecated without telling what to use instead?

**Comment by nithinsubbiah** [comment]
> The lastest API doc marks the majority of the context management APIs as [deprecated](https://rocm.docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___context_d.html#ga8aa32cf64272da929f23ecbafefefcee), with some exceptions like `hipDevicePrimaryCtxRetain`. Not sure that's intentional or just a bug in the doc. Could you ask internally to clarify regarding this aspect? And also regarding primary context management (is it really needed with `hipDevicePrimaryCtxRelease` doing nothing?) And what is the expected way to manage context in general, with a bunch of APIs deprecated without telling what to use instead?

You're right that most context API's are deprecated except a handful like `hipDevicePrimaryCtxRetain`, `hipDevicePrimaryCtxRelease` (https://docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___context.html)

As for the reasoning and ways to manage context going forward, the documentation recommends using stream and device APIs. 

Screenshot 2023-12-12 at 11 33 49 AM

**Comment by benvanik** [comment]
yeah we don't want the runtime API - it adds extra baggage - sounds like they just want people to use that in HIP land vs actually having deprecated things?

**Comment by nithinsubbiah** [comment]
Yeah, that's my understanding. I had to use `hipSetDevice` (https://docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___device.html#ga43c1e7f15925eeb762195ccb5e063eae) from the runtime API to steer clear of the deprecated Ctx API

**Comment by benvanik** [comment]
yeah we don't want that, so we'll want to go back to the context ones (they can never remove those, so it's not deprecated in the sense of "don't use these because they won't work/will be removed)

**Comment by nithinsubbiah** [comment]
I've replaced the device call with context. I'll follow up internally on these "deprecated" context API calls to make sure they're supported in future releases

**Comment by antiagainst** [comment]
> yeah we don't want the runtime API - it adds extra baggage - sounds like they just want people to use that in HIP land vs actually having deprecated things?

[Per the doc](https://rocm.docs.amd.com/projects/HIP/en/docs-5.6.0/user_guide/hip_porting_driver_api.html#hip-module-and-ctx-apis), in HIP there is no clear separation of driver vs runtime APIs. So I suspect we might be seeing the runtime API (in CUDA's sense) overhead in HIP anyway. From the doc it seems `Ctx` APIs are just sugar over other APIs, but there are always suprises. :) Worth digging into the details and pick folks' brains internally to understand in depth. @nithinsubbiah please make sure this is tracked and investigated.

File: experimental/hip/dynamic_symbol_tables.h [hunk 6]
```diff
@@ -62,6 +63,7 @@ IREE_HIP_PFN_DECL(hipModuleLoadData, hipModule_t *, const void *)
 IREE_HIP_PFN_DECL(hipModuleLoadDataEx, hipModule_t *, const void *,
                   unsigned int, hipJitOption *, void **)
 IREE_HIP_PFN_DECL(hipModuleUnload, hipModule_t)
+IREE_HIP_PFN_DECL(hipSetDevice, int)
```
Comment (antiagainst) [inline experimental/hip/dynamic_symbol_tables.h:66]: This is not used and can be removed now?

**Review by antiagainst — APPROVED** [review]

**Pull request merged**